src/main/java/com/jensjansson/ce/bot/BotManager.java

```diff
@@ -0,0 +1,347 @@
+import com.vaadin.collaborationengine.TopicConnection;
+import com.vaadin.collaborationengine.UserInfo;
+import com.vaadin.flow.shared.Registration;
+
+import com.jensjansson.ce.data.entity.Person;
+import com.jensjansson.ce.data.service.PersonService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Thread that controls the bots, both for presence and for editing forms.
+ */
+public class BotManager implements Runnable {
+    public static final String BOT_PREFIX =
+        "pr-" + BotUserGenerator.BOT_ID_PREFIX;
+    private final Random random = new Random();
+    private static final Logger logger = LoggerFactory
+        .getLogger(BotManager.class);
+
+    private static final int botCount = 20;
+    private static final int maxAvatarNumber = 8;
+    private static BotManager instance;
+
+    private PersonService personService;
+    private CollaborationEngine ce;
+    private List<Integer> ids = Collections.emptyList();
+    private List<UserInfo> bots;
+    private List<ExtraBot> extraBots;
+
+    private BiFunction<UserInfo, String, PresenceManager> presenceManagerCreator;
+    private Map<Integer,UserHandler> handlerMap;
+    private ConcurrentHashMap<Integer, Bot> botMap = new ConcurrentHashMap<>();
+
+    BotManager(PersonService personService, CollaborationEngine ce) {
+        this.personService = personService;
+        this.ce = ce;
+        this.bots = createBotUsers(botCount);
+        this.extraBots = createBotUsers(5).stream().map(ExtraBot::new).collect(
+            Collectors.toList());
+        this.presenceManagerCreator = createPresenceManagerCreator();
+    }
+
+    public static void createInstance(PersonService personService,
+        CollaborationEngine ce) {
+        if (instance != null) {
+            throw new IllegalStateException(
+                "Only 1 instance should be created");
+        }
+        instance = new BotManager(personService, ce);
+        Thread thread = new Thread(instance);
+        thread.setDaemon(true);
+        thread.setName("Bot-Thread");
+        thread.start();
+    }
+
+    private int generateNumberOfEditsBeforeSave() {
+        return 2 + (random.nextInt(4));
+    }
+
+    @Override
+    public synchronized void run() {
+        boolean initialized = false;
+        while (!initialized) {
+            doWait(1000);
+            initialized = initializeIfPossible();
+        }
+        IntStream.range(0,extraBots.size()).forEach(this::createPresence);
+        for (int i = 0; true; i = (i + 1) % extraBots.size()) {
+            if(random.nextDouble() < 0.2) {
+                createPresence(i);
+            }
+            UserHandler handler = handlerMap.get(ids.get(random.nextInt(ids.size())));
+            //Ignore if has users.
+            if(handler.hasUsers) continue;
+            handler.presenceManager.markAsPresent(random.nextBoolean());
+
+            doWait(500);
+
+            logger.debug("Running bots");
```

Comment:
This logs a lot since Spring if configured with `trace` logging, I guess it was more useful during development.

Reply:
Yes, you are right. I removed the line.